File: grpc-pg-bench/spring-vt/src/main/java/com/beam/bench/GrpcServerConfig.java
package com.beam.bench;

import io.grpc.netty.NettyServerBuilder;
import io.netty.channel.ChannelOption;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.MultiThreadIoEventLoopGroup;
import io.netty.channel.epoll.Epoll;
import io.netty.channel.epoll.EpollIoHandler;
import io.netty.channel.epoll.EpollServerSocketChannel;
import io.netty.channel.nio.NioIoHandler;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import jakarta.annotation.PreDestroy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.grpc.server.ServerBuilderCustomizer;

import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * Tunes the grpc-netty server for this 2-core box. Two things matter here, both
 * justified by the CPU profile (≈72% of CPU was in the Netty/gRPC transport,
 * ≈57% in syscalls + cross-thread wakeups — see spring-vt/PROFILING.md):
 *
 *  1. <b>Run RPC handlers on virtual threads.</b> Without an explicit executor,
 *     grpc-java uses its default cached <i>platform</i> pool; setting the VT
 *     executor is what makes each handler — and the blocking JDBC inside it —
 *     park a virtual thread on I/O. (This was the original spring-vt behaviour.)
 *
 *  2. <b>Use the Linux epoll transport instead of NIO.</b> epoll avoids the NIO
 *     selector's per-loop syscall overhead and is the standard Linux gRPC tuning.
 *     Netty 4.2 deprecated {@code EpollEventLoopGroup}; the current API is
 *     {@link MultiThreadIoEventLoopGroup} + {@link EpollIoHandler}. We fall back
 *     to NIO automatically if epoll isn't available (e.g. on macOS).
 *
 * Plus small HTTP/2 knobs matching the go-pgx server: 1 MiB flow-control window,
 * TCP_NODELAY (gRPC writes are tiny), and server keepalive. The event-loop
 * threads do I/O only (handlers are off-loaded to VTs), so we size the worker
 * group to the available cores rather than Netty's default 2×cores.
 */
@Configuration
class GrpcServerConfig {

    private static final Logger log = LoggerFactory.getLogger(GrpcServerConfig.class);

    // Netty I/O (worker) event-loop thread count — the single source of truth is
    // application.properties (NETTY_IO_THREADS=1). Spring resolves this placeholder
    // from the OS environment FIRST and only then from application.properties, so
    // the perf A/B harness can still force 2 with `NETTY_IO_THREADS=2` in the env
    // without editing the file. 0/blank => auto = cores/2 (1 under the taskset pin).
    // This sizes the WORKER group only; the boss/accept group is always 1.
    @Value("${NETTY_IO_THREADS:0}")
    private int configuredIoThreads;

    private EventLoopGroup boss;
    private EventLoopGroup worker;

    @Bean
    ServerBuilderCustomizer<NettyServerBuilder> tunedNettyServerExecutor() {
        // I/O-thread count comes from application.properties (NETTY_IO_THREADS=1),
        // injected above; 0 means auto = cores/2. availableProcessors() honours the
        // taskset CPU affinity on Linux, so cores/2 is 1 on the 2-core-pinned box.
        // Why 1: the handlers run on virtual threads, so the Netty event loop only
        // does socket I/O + HTTP/2 framing. An A/B on the 2-core box showed 1 I/O
        // thread beats 2 by up to +14% (and cuts p99 ~30% at c=128). MECHANISM
        // (perf-stat A/B, results/perf-cohost2-*): it is NOT fewer context switches —
        // those stayed ~flat (~18k/s, ~6-7 per million instructions) whether 1 or 2
        // I/O threads, because ctx switches are driven by virtual-thread park/unpark
        // on the blocking JDBC round-trips, not by the event loops. The real cost of
        // a 2nd I/O thread is CACHE LOCALITY: two event loops split across cores 2,3
        // bounce connection/buffer state between the two cores' caches → IPC -6%
        // (0.483→0.454) and cache-misses +10%. On this DB-bound path rps looks
        // identical (Postgres is the wall), but 1 I/O thread is the more CPU-efficient
        // choice. cores/2 → 1 on 2 cores, 2 on 4 cores.
        int ioThreads = configuredIoThreads > 0
                ? configuredIoThreads
                : Math.max(1, Runtime.getRuntime().availableProcessors() / 2);
        // NETTY_TRANSPORT=nio forces NIO even on Linux — lets us A/B the same
        // binary (epoll vs nio) back-to-back in one session to control for
        // day-to-day box variance. Default (unset/"epoll") uses epoll if available.
        boolean forceNio = "nio".equalsIgnoreCase(System.getenv("NETTY_TRANSPORT"));
        boolean epoll = !forceNio && Epoll.isAvailable();
        if (epoll) {
            boss = new MultiThreadIoEventLoopGroup(1, EpollIoHandler.newFactory());
            worker = new MultiThreadIoEventLoopGroup(ioThreads, EpollIoHandler.newFactory());
        } else {
            boss = new MultiThreadIoEventLoopGroup(1, NioIoHandler.newFactory());
            worker = new MultiThreadIoEventLoopGroup(ioThreads, NioIoHandler.newFactory());
        }
        log.info("grpc-netty transport: {} (io threads={})", epoll ? "epoll" : "nio", ioThreads);

        final EventLoopGroup bossRef = boss;
        final EventLoopGroup workerRef = worker;
        return builder -> builder
                .executor(Executors.newVirtualThreadPerTaskExecutor())
                .channelType(epoll ? EpollServerSocketChannel.class : NioServerSocketChannel.class)
                .bossEventLoopGroup(bossRef)
                .workerEventLoopGroup(workerRef)
                .withChildOption(ChannelOption.TCP_NODELAY, true)
                .flowControlWindow(1 << 20)
                .maxConcurrentCallsPerConnection(Integer.MAX_VALUE)
                .keepAliveTime(30, TimeUnit.SECONDS)
                .keepAliveTimeout(10, TimeUnit.SECONDS)
                .permitKeepAliveWithoutCalls(true);
    }

    @PreDestroy
    void shutdownEventLoops() {
        if (worker != null) {
            worker.shutdownGracefully();
        }
        if (boss != null) {
            boss.shutdownGracefully();
        }
    }
}

File: grpc-pg-bench/spring-rt/src/main/kotlin/com/beam/bench/GrpcServerConfig.kt
package com.beam.bench

import io.grpc.netty.NettyServerBuilder
import io.netty.channel.ChannelOption
import io.netty.channel.EventLoopGroup
import io.netty.channel.MultiThreadIoEventLoopGroup
import io.netty.channel.epoll.Epoll
import io.netty.channel.epoll.EpollIoHandler
import io.netty.channel.epoll.EpollServerSocketChannel
import io.netty.channel.nio.NioIoHandler
import io.netty.channel.socket.nio.NioServerSocketChannel
import jakarta.annotation.PreDestroy
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.grpc.server.ServerBuilderCustomizer
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit

/**
 * Netty tuning for the **reactive** spring-rt server on the 2-core box. The
 * baseline wall-clock profile showed ~96% of time in syscalls + cross-thread
 * wakeups (`pthread_cond_signal`/`__lll_lock_wait`) — i.e. the cost is thread
 * hops and per-loop syscalls, NOT compute. So we cut both:
 *
 *  1. **Run handlers inline on the Netty event loop (direct executor).** Unlike
 *     spring-vt — which sets a *virtual-thread* executor so blocking JDBC can
 *     park a carrier — spring-rt is fully non-blocking (coroutines + R2DBC), so
 *     a VT executor would be exactly wrong: it would add a pointless thread hop
 *     per call. A **direct executor** instead starts each coroutine on the I/O
 *     thread that decoded the request; it only leaves that thread when it
 *     suspends on R2DBC (no work is blocked on the loop). This removes the
 *     hand-off to grpc-java's default platform pool — one fewer wakeup per RPC.
 *     SAFE ONLY because nothing on this path blocks (verified via `wall` profile).
 *
 *  2. **Linux epoll instead of NIO.** Fewer selector syscalls per loop than the
 *     NIO selector; the standard Linux gRPC tuning. Falls back to NIO if epoll
 *     is unavailable.
 *
 * Plus the same HTTP/2 knobs as go-pgx / spring-vt: 1 MiB flow-control window,
 * TCP_NODELAY (tiny gRPC writes), server keepalive. I/O threads default to the
 * visible core count (handlers run ON the loop here, so unlike spring-vt we do
 * NOT shrink to cores/2); override with NETTY_IO_THREADS to A/B.
 */
@Configuration
class GrpcServerConfig {

    private val log = LoggerFactory.getLogger(GrpcServerConfig::class.java)

    private var boss: EventLoopGroup? = null
    private var worker: EventLoopGroup? = null

    /** Executor that runs each task inline on the caller (the Netty I/O thread). */
    private val directExecutor = Executor { it.run() }

    @Bean
    fun reactiveNettyServerCustomizer(): ServerBuilderCustomizer<NettyServerBuilder> {
        // GRPC_TUNED=off returns the builder untouched (default NIO transport +
        // grpc-java's default platform executor) — the un-tuned baseline, for a
        // rigorous same-build A/B against the tuned path.
        if ("off".equals(System.getenv("GRPC_TUNED"), ignoreCase = true)) {
            log.info("grpc-netty: tuning DISABLED (GRPC_TUNED=off) — default NIO + default executor")
            return ServerBuilderCustomizer { }
        }

        // Handlers run ON the event loop (direct executor) only until they suspend
        // on R2DBC, so — like spring-vt — fewer loops means fewer cross-thread
        // wakeups. Default to cores/2 (=1 on the 2-core box); the A/B sweep showed
        // pool>32 and extra I/O threads add contention, not throughput.
        val ioEnv = System.getenv("NETTY_IO_THREADS")
        val ioThreads = if (!ioEnv.isNullOrBlank()) ioEnv.trim().toInt()
        else maxOf(1, Runtime.getRuntime().availableProcessors() / 2)

        val forceNio = "nio".equals(System.getenv("NETTY_TRANSPORT"), ignoreCase = true)
        val epoll = !forceNio && Epoll.isAvailable()
        if (epoll) {
            boss = MultiThreadIoEventLoopGroup(1, EpollIoHandler.newFactory())
            worker = MultiThreadIoEventLoopGroup(ioThreads, EpollIoHandler.newFactory())
        } else {
            boss = MultiThreadIoEventLoopGroup(1, NioIoHandler.newFactory())
            worker = MultiThreadIoEventLoopGroup(ioThreads, NioIoHandler.newFactory())
        }
        log.info("grpc-netty transport: {} (io threads={}, direct executor)", if (epoll) "epoll" else "nio", ioThreads)

        val bossRef = boss!!
        val workerRef = worker!!
        return ServerBuilderCustomizer { builder ->
            builder
                .executor(directExecutor)
                .channelType(if (epoll) EpollServerSocketChannel::class.java else NioServerSocketChannel::class.java)
                .bossEventLoopGroup(bossRef)
                .workerEventLoopGroup(workerRef)
                .withChildOption(ChannelOption.TCP_NODELAY, true)
                .flowControlWindow(1 shl 20)
                .maxConcurrentCallsPerConnection(Int.MAX_VALUE)
                .keepAliveTime(30, TimeUnit.SECONDS)
                .keepAliveTimeout(10, TimeUnit.SECONDS)
                .permitKeepAliveWithoutCalls(true)
        }
    }

    @PreDestroy
    fun shutdownEventLoops() {
        worker?.shutdownGracefully()
        boss?.shutdownGracefully()
    }
}
